crates/holochain/src/core/workflow/sys_validation_workflow.rs

```diff
@@ -418,77 +419,99 @@ fn get_dependency_hashes_from_ops(ops: impl Iterator<Item = DhtOpHashed>) -> Vec
-                                    entry,
-                                    session_data,
-                                )
-                                .unwrap_or_else(|_| vec![])
-                                .into_iter()
-                                .map(|action| action.into_hash())
-                                .collect::<Vec<_>>()
+                            if let NewEntryAction::Update(update) = action {
+                                actions.push(update.original_action_address.clone());
                             }
-                            _ => vec![],
-                        };
-
-                        if let NewEntryAction::Update(update) = action {
-                            actions.push(update.original_action_address.clone());
+                            Some(actions)
                         }
-                        Some(actions)
-                    }
-                    ChainOp::RegisterAgentActivity(_, action) => action
-                        .prev_action()
-                        .map(|action| vec![action.as_hash().clone()]),
-                    ChainOp::RegisterUpdatedContent(_, action, _) => {
-                        Some(vec![action.original_action_address.clone()])
-                    }
-                    ChainOp::RegisterUpdatedRecord(_, action, _) => {
-                        Some(vec![action.original_action_address.clone()])
-                    }
-                    ChainOp::RegisterDeletedBy(_, action) => {
-                        Some(vec![action.deletes_address.clone()])
+                        ChainOp::RegisterAgentActivity(_, action) => action
+                            .prev_action()
+                            .map(|action| vec![action.as_hash().clone()]),
+                        ChainOp::RegisterUpdatedContent(_, action, _) => {
+                            Some(vec![action.original_action_address.clone()])
+                        }
+                        ChainOp::RegisterUpdatedRecord(_, action, _) => {
+                            Some(vec![action.original_action_address.clone()])
+                        }
+                        ChainOp::RegisterDeletedBy(_, action) => {
+                            Some(vec![action.deletes_address.clone()])
+                        }
+                        ChainOp::RegisterDeletedEntryAction(_, action) => {
+                            Some(vec![action.deletes_address.clone()])
+                        }
+                        ChainOp::RegisterRemoveLink(_, action) => {
+                            Some(vec![action.link_add_address.clone()])
+                        }
+                        _ => None,
+                    };
+
+                    // Use hash set to prevent duplicate hashes.
+                    let mut dependency_hashes = HashSet::new();
+                    // If a previous action exists, it needs to be fetched to check agent validity, independent of the op type.
```

Comment:
Not clear why we need to do this. This means we end up fetching the entire chain for every agent that publishes an op which fits into our arc. That effectively prevents sharding.

I think we should only be responsible for checking that the agent key is valid when we are an agent activity authority. In which case we are already getting the previous action

Reply:
We need to do this to check if the previous action deleted the agent key.
No, we're not fetching the entire chain, we're only fetching the previous action of each op we're validating. That previous action is not validated and nothing else is fetched.

> I think we should only be responsible for checking that the agent key is valid when we are an agent activity authority. In which case we are already getting the previous action

Okay, I'll update that.